File: code/classes/downcasting.m
#import <Foundation/Foundation.h>

// Downcasting in Objective-C
@interface Animal : NSObject
- (void)makeSound;
@end

@implementation Animal
- (void)makeSound {
    NSLog(@"Some sound");
}
@end

@interface Dog : Animal
- (void)fetch;
@end

@implementation Dog
- (void)makeSound {
    NSLog(@"Woof");
}
- (void)fetch {
    NSLog(@"Fetching...");
}
@end

int main() {
    Animal *animal = [[Dog alloc] init];
    
    // Downcasting with type checking
    if ([animal isKindOfClass:[Dog class]]) {
        Dog *dog = (Dog *)animal;
        [dog makeSound];
        [dog fetch];
    }
    
    return 0;
}

File: code/classes/downcasting.ex
defmodule Animal do
  defstruct [:type]
  
  def make_sound(%{type: :dog}), do: "Woof"
  def make_sound(_), do: "Some sound"
end

defmodule Dog do
  defstruct [:type, :breed]
  
  def fetch(%{breed: breed}), do: "Fetching #{breed} style"
end

# Create a dog
dog = %Dog{type: :dog, breed: "Golden Retriever"}

# Pattern match to simulate downcasting
case dog do
  %Dog{breed: breed} -&gt; 
    IO.puts Animal.make_sound(dog)
    IO.puts Dog.fetch(dog)
  _ -&gt; 
    IO.puts Animal.make_sound(dog)
end